src/ucp/core/ucp_rkey.c

```diff
@@ -213,61 +218,507 @@ UCS_PROFILE_FUNC(ssize_t, ucp_rkey_pack_memh,
                                 sys_dev_map, sys_distance, buffer, 1, 0);
 }
 
+static size_t ucp_memh_extended_info_size(size_t common_size,
+                                          size_t specific_info_size)
+{
+    if ((specific_info_size + common_size - sizeof(uint8_t)) > UINT8_MAX) {
+        /* This field is packed after the 8-bit size field in case of a total
+         * info size is greater than UINT8_MAX. */
+        return sizeof(uint16_t);
+    }
+
+    return 0;
+}
+
+static size_t ucp_memh_common_packed_size(size_t specific_info_size)
+{
+    size_t size;
+
+    /* Size of mkey information which comes prior TL mkey data of all MDs */
+    size = sizeof(uint8_t);
+
+    /* Flags */
+    size += sizeof(uint16_t);
+
+    /* Memory domains map */
+    size += sizeof(uint64_t);
+
+    /* Memory type */
+    size += sizeof(uint8_t);
+
+    return size + ucp_memh_extended_info_size(size, specific_info_size);
+}
+
+static size_t ucp_memh_exported_info_packed_size()
+{
+    size_t size;
+
+    /* Address */
+    size = sizeof(uint64_t);
+
+    /* Length */
+    size += sizeof(uint64_t);
+
+    /* UCP uuid */
+    size += sizeof(uint64_t);
+
+    /* Registration ID */
+    size += sizeof(uint64_t);
+
+    /* Common data (size, md_map, mem_type and etc) */
+    size += ucp_memh_common_packed_size(size);
+
+    return size;
+}
+
+static void ucp_memh_info_size_pack(void **p, size_t info_size)
+{
+    ucs_assert(info_size <= UINT16_MAX);
+
+    if ((info_size - sizeof(uint8_t)) > UINT8_MAX) {
+        *ucs_serialize_next(p, uint8_t)  = 0;
+        *ucs_serialize_next(p, uint16_t) = info_size - sizeof(uint16_t);
```

Comment:
pls add assert that it is less than `UINT16_MAX`

Reply:
see line 276 - I guess it should be enough